### ISSUE TITLE
Issue using graphics::clear

### ISSUE THREAD

**Issue by KyleGBC** [ggez/ggez#921]
I'm just started with using this library and ran into an issue with my project almost immediately. I was able to replicate my problem with a very simple program, leading me to believe it is a result of my own misunderstanding of the library. 

The basic program I set up to help me demonstrate is as follows: 
```
use ggez::{graphics, Context, ContextBuilder, GameResult};
use ggez::event::{self, EventHandler, KeyCode, KeyMods};

fn main() {
    let (mut ctx, mut event_loop) = ContextBuilder::new("Game", "Kyle").build().expect(">:(");
    let mut my_game = MyGame::new(&mut ctx);

    match event::run(&mut ctx, &mut event_loop, &mut my_game) {
        Ok(_) => println!("Exited cleanly."),
        Err(e) => println!("Error occured: {}", e)
    }
}

struct MyGame {
    image: graphics::Image,
    x: f32,
    y: f32,
}

impl MyGame {
    pub fn new(_ctx: &mut Context) -> MyGame {
        let image = graphics::Image::new(_ctx, "/red.png").unwrap();
        MyGame {
		    image, x: 10.0, y: 10.0
		}
    }
}

impl EventHandler for MyGame {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
		Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
	graphics::clear(ctx, graphics::WHITE);

        let drawparams = graphics::DrawParam::default().dest([self.x, self.y]);
        graphics::draw(ctx, &self.image, drawparams)?;

        println!("Drawing at {}, {}", self.x, self.y);

	graphics::present(ctx)
    }
    fn key_down_event(&mut self, _ctx: &mut Context, keycode: KeyCode, _keymods: KeyMods, _repeat: bool){
        match keycode {
            KeyCode::Up => self.y -= 1.0,
            KeyCode::Down => self.y += 1.0,
            KeyCode::Right => self.x += 1.0,
            KeyCode::Left => self.x -= 1.0,
            _ => ()
        }
    }
}
```

As you can see, all this does is load an image, in this case, a 1x1 red png, and draw it at the stored coordinates, which are changed by pressing the arrow keys. 

However, running this doesn't work, and the image does not move around with the keypresses (although the `println!()` within the `MyGame::draw function` _does_ change accordingly).

When the `graphics::clear(ctx, graphics::WHITE)` is removed from `My::Game::draw`, the image _*does*_ move around with the keypresses, although obviously "burned in" so to speak as nothing is clearing out the red image from the previous locations. 

I'm doing something wrong here to create this issue but for the life of me, cannot say what. Any help appreciated. (I can make a video of this code snippet running with and without the `graphics::clear` if that makes the problem more apparent)

Update: 
I tried running the `04_snake.rs` example code from this repo and observed the same behavior as with my minimal project.
Perhaps the issue is with my environment ¯\\_ (ツ)_/¯

**Comment by PSteinhaus** [comment]
I just tested your code and it runs just as one would expect on my machine (Windows 10, ggez 0.5.1). The 1 px square is shown and moves on keypress. When I remove the call to ```clear``` then I see the burned in effect, just as expected.

What is your environment if I may ask?

**Comment by KyleGBC** [comment]
> I just tested your code and it runs just as one would expect on my machine (Windows 10, ggez 0.5.1). The 1 px square is shown and moves on keypress. When I remove the call to `clear` then I see the burned in effect, just as expected.
> 
> What is your environment if I may ask?

Not entirely sure all what info is relevant here but I'll attempt.

- rustc/cargo version: 1.50.0
- ggez version: 0.5.1
- toolchain: stable-x86_64-pc-windows-msvc
- Windows 10.0.19043
- On a laptop with both integrated graphics (Intel) and discrete graphics (NVidia).

**Comment by PSteinhaus** [comment]
Hmm, seems pretty similar to my setup.
What's your output when you run ```graphics::renderer_info``` like this:
```rust
println!("Renderer: {}", graphics::renderer_info(ctx).unwrap());
```
I'm sadly no expert on this territory, but my first guess would of course be that it's caused by the driver not reacting to ```gfx::Encoder::clear_raw``` as expected, which would mean that it's probably a dependency bug in gfx...

Could you test whether your problem persists when you use the current devel branch of ggez? Perhaps the bug already got fixed somewhere in the dependencies.

**Comment by KyleGBC** [comment]
@PSteinhaus 
Interesting(?)

The `graphics::render_info` `println!` prints the following:
`Renderer: Requested OpenGl 3.2 Core profile, actually got Driver vendor: Intel, renderer Intel(R) UHD Graphics 630, version 3.2.0, - Build 30.0.100.9458, shading language 1.50, - Build 30.0.100.9458.`

Not too sure if that means anything but there you go.

As for the dev branch, it doesn't build due to a missing comma in `ggez::graphics::canvas::with_window_size()` that was fixed 12 days ago in a pull request that hasn't been accepted yet.

**Comment by PSteinhaus** [comment]
Well now we know at least that it's the Intel driver that is involved. So changing the driver might enable you to use ggez, though it of course wouldn't solve the underlying problem...

If the dev branch doesn't build, but the pull request does, then just build the latter.

To be honest this is interesting, but I'm probably not able to help you.
The best we might be able to do is test more graphics functions to see if we can pin it down to a specific issue inside gfx. Then we could report it and hope it gets fixed some time in the future...

**Comment by PSteinhaus** [comment]
Now that the PR fixing the build error has been merged could you try again to see if the problem still persists in the current devel branch?

**Comment by PSteinhaus** [comment]
@KyleGBC could you please check if your issue still persists in the latest devel branch?
If not I just have to assume that the issue no longer persists and close this.

**Comment by junkmail22** [comment]
Don't close this issue just yet!

I've managed to find this error in action as well! It has nothing to do with the development environment, and has to do with where the program is run. In other words, once I compile and distribute the executable, the error does not appear on my computer, but it consistently appears on one of my tester's computers.

I've tested this with a small test program on devel branch: 

`use ggez::{Context, ContextBuilder, GameResult};
use ggez::event::{self, EventHandler};
use ggez::graphics;
use ggez::conf::WindowMode;
use ggez::conf::WindowSetup;

use mint;

type Point2 = mint::Point2<f32>;

const DEFAULT_WINDOW_HEIGHT: f32 = 960f32;
const DEFAULT_WINDOW_WIDTH: f32 = 1440f32;

fn main() {
    // Make a Context and an EventLoop.
    let (mut ctx,mut event_loop) = ContextBuilder::new("Mouse Test!", "junkmail")
        .window_setup(WindowSetup::default().title("Mouse Test!"))
        .window_mode(
            WindowMode::default().dimensions(DEFAULT_WINDOW_WIDTH, DEFAULT_WINDOW_HEIGHT),
        )
        .build()
        .unwrap();

    let rect = graphics::Rect::new(0.0, 0.0, DEFAULT_WINDOW_WIDTH, DEFAULT_WINDOW_HEIGHT);

    graphics::set_screen_coordinates(&mut ctx, rect);
    let mut my_game = MyGame::new(&mut ctx);

    // Run!
    event::run( ctx, event_loop, my_game);
}

struct MyGame {
    image:graphics::Image,
    mouse_position:(f32,f32),
}

impl MyGame {
    pub fn new(ctx: &mut Context) -> MyGame {
        // Load/create resources here: images, fonts, sounds, etc.
        MyGame
        {
            image:graphics::Image::new(ctx, "/crosshair.png").unwrap(),
            mouse_position:(0.0, 0.0),
        }
    }
}

impl EventHandler for MyGame {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        // Update code here...
        Ok(())
    }

    fn mouse_motion_event(&mut self, _ctx: &mut Context, x: f32, y: f32, _xrel: f32, _yrel: f32) {
        self.mouse_position = (x,y);
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx, graphics::Color::BLACK);//removing this line of code causes the image to update, but not wipe the screen

        let (x,y) = self.mouse_position;

        let pos = Point2{x: x-24.0, y: y-24.0};
        let drawparams = graphics::DrawParam::new().dest(pos);
        graphics::draw(ctx, &self.image, drawparams).unwrap();

        graphics::present(ctx)
    }
}
`

The error occurs exactly as @KyleGBC described. Without the graphics::clear call, the image does update (but smears). With it, the display fails to update. It's not just a minimal example: it also occurs in my full game project.

This issue is utterly cursed. I might be able to provide more specs on request, but I can't quickly reproduce and test changes since the error does not appear when programs are run on my machine.

**Comment by PSteinhaus** [comment]
@junkmail22 Hey, thanks for that info!
Your stock of testers comes in handy again and again ^^
Then this will have to stay open (sadly).

I just don't know what may cause this issue comes. All I can imagine right now is, as I already said, that this is a dependency bug on gfx's side with `gfx::Encoder::clear_raw` not working properly, [since calling this is all that `graphics::clear` does, really](https://github.com/ggez/ggez/blob/7f117f8e95393dd27353d9adf16edfbfdef5c353/src/graphics/mod.rs#L452-L458).

`gfx::Encoder::clear_raw` actually calls `gfx_device_gl::CommandBuffer::clear_color`(https://docs.rs/gfx_device_gl/0.16.2/gfx_device_gl/struct.CommandBuffer.html#method.clear_color), which seems to do mainly two things:
- create a "clear"-variant of a command enum and push it to the buffer of commands
- and create a `PixelTargetSet` with the color target view and "bind" to it

The second seems to be where the magic happens, so I assume it's where a bug might be found, especially since `CommandBuffer::bind_pixel_targets` only seems to be used in the two clear functions (as far as I can see right now).

**Comment by PSteinhaus** [comment]
> I might be able to provide more specs on request

That would probably be useful indeed. Asking the specific tester to run [this](https://github.com/ggez/ggez/issues/921#issuecomment-815616713) may (or may not) give us a clue for example.

**Comment by PSteinhaus** [comment]
One thing that anyone experiencing this may try as well is changing the graphics backend to GLES, [as shown in the `graphics_settings` example](https://github.com/ggez/ggez/blob/359442b7869b580e8dae8435226e3888617ea5e8/examples/graphics_settings.rs#L215-L232).
This is a shot in the dark, but if that worked we'd at least have a workaround...

Another idea would be to put the `clear` call right before `present` to see whether this changes the behaviour. If so, then maybe the interpretation of the clear variant of the `Command` enum is flawed, so that all graphics commands until the first appearance of _clear_ are handled correctly and then it fails silently. But then again, probably it shouldn't change after all, since the finalizing steps like the draw buffer swap shouldn't execute then as well...

**Comment by junkmail22** [comment]
Here's the DxDiag from the computer which has the issue. I'll try builds with the changes suggested soon as well.

[DxDiag_Clear_Failure.txt](https://github.com/ggez/ggez/files/6695998/DxDiag_Clear_Failure.txt)

**Comment by KyleGBC** [comment]
> Here's the DxDiag from the computer which has the issue. I'll try builds with the changes suggested soon as well.
> 
> [DxDiag_Clear_Failure.txt](https://github.com/ggez/ggez/files/6695998/DxDiag_Clear_Failure.txt)

I've moved on from this library a long time ago but I would be remiss not to mention that the machine you are experiencing the issue on is actually the same model as the one I experienced it on. Here is my [DxDiag](https://github.com/ggez/ggez/files/6697453/DxDiag_KyleGBC.txt).

**Comment by PSteinhaus** [comment]
Thanks for the DxDiags!

So it seems that it may really be a hardware/driver issue. Too bad that it happens with both the older and the (I guess) up-to-date driver for that Intel UHD graphics chip...

On second thought, what does seem strange is that ggez (or more precisely glutin) chooses the onboard graphics instead of the way beefier GPUs that both of these PCs have... and I don't see why or how to change that.

**Comment by junkmail22** [comment]
It looks like setting the preferred graphics device to the graphics card worked! Resolved the issue on my tester's end. Not sure how to proceed, but it did fix the issue in this case.

**Comment by PSteinhaus** [comment]
Thanks for the feedback!
Then we at least have a workaround for this problem for anyone who has a GPU in addition of the onboard graphics. That's not everyone, but it's better than nothing.

It seems like we cannot _choose_ the graphics device programatically, so we pretty much have to rely on the end-users fixing this problem for themselves, sadly.

I'll label this "won't fix", since it's a dependency bug in gfx pre-ll, which is no longer actively maintained.

**Comment by icefoxen** [comment]
Some history in the issue of the "tell program which GPU to use" problem is here: https://github.com/ggez/ggez/issues/186 and in related issues.

TLDR; OpenGL and the surrounding standards don't actually specify a way to look at the GPU's available and say "I want that one".  It's semi-possible on a few systems, but it's platform-specific and/or done via vendor extensions.  All in all, horrible.  Adding a standard way to do something so simple and pragmatic is one of my favorite things about Vulkan.

**Comment by PSteinhaus** [comment]
With thw wgpu-transition done Vulkan is now the preferred backend and wgpu also allows you to [choose your adapter](https://github.com/gfx-rs/wgpu#environment-variables).
